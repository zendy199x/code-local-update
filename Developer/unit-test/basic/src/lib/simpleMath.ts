export const sum = (a: number, b: number): number => {
  if (a == null || b == null) {
    return null;
  }
  return a + b;
}

export const prod = (a: number, b: number): Promise<Number> => {
  if (a == null || b == null) {
    return Promise.reject('Me not do null prods');
  }
  return Promise.resolve(a * b);
}

export const divide = (a: number, b: number): number => {
  if (a == null || b == null) {
    return null;
  } else if (!b) {
    return Number.POSITIVE_INFINITY;
  }
  return a / b;
}