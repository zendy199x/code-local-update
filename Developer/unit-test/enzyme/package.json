{
  "name": "samples-for-react-jest-enzyme-ts-unit-test-training",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "start": "webpack-dev-server",
    "lint": "tslint -c ./tslint.json --project ./tsconfig.json \"src/**/*.{ts,tsx}\" -e \"**/eds/**\"",
    "test": "jest --verbose",
    "test:watch": "jest --watchAll --verbose --i",
    "test:coverage": "jest --verbose --coverage",
    "test:update_snapshots": "jest --verbose -u"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/arp82/samples-for-react-jest-enzyme-ts-unit-test-training.git"
  },
  "author": "arp82",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/arp82/samples-for-react-jest-enzyme-ts-unit-test-training/issues"
  },
  "homepage": "https://github.com/arp82/samples-for-react-jest-enzyme-ts-unit-test-training#readme",
  "devDependencies": {
    "@types/deep-freeze": "0.1.1",
    "@types/enzyme": "3.1.10",
    "@types/jest": "22.2.3",
    "@types/lodash-es": "4.17.0",
    "@types/react": "16.4.1",
    "@types/react-dom": "16.0.6",
    "@types/react-redux": "6.0.2",
    "@types/react-router": "3.0.15",
    "@types/redux-mock-store": "0.0.13",
    "awesome-typescript-loader": "3.5.0",
    "axios-mock-adapter": "1.15.0",
    "babel-core": "6.26.3",
    "babel-preset-env": "1.7.0",
    "cross-env": "5.2.0",
    "css-loader": "0.28.11",
    "deep-freeze": "0.0.1",
    "enzyme": "3.3.0",
    "enzyme-adapter-react-16": "1.1.1",
    "enzyme-to-json": "3.3.4",
    "extract-text-webpack-plugin": "3.0.2",
    "file-loader": "1.1.11",
    "html-webpack-plugin": "2.30.1",
    "identity-obj-proxy": "3.0.0",
    "jest": "22.4.4",
    "node-sass": "4.9.0",
    "redux-mock-store": "1.4.0",
    "sass-loader": "6.0.7",
    "speed-measure-webpack-plugin": "1.2.2",
    "style-loader": "0.19.1",
    "ts-jest": "22.4.6",
    "tslint": "5.10.0",
    "tslint-react": "3.6.0",
    "typescript": "2.9.2",
    "webpack": "3.12.0",
    "webpack-bundle-analyzer": "2.13.1",
    "webpack-dev-server": "2.11.2",
    "webpack-merge": "4.1.3"
  },
  "dependencies": {
    "@material-ui/core": "1.4.3",
    "@material-ui/icons": "1.1.0",
    "@turf/circle": "6.0.1",
    "axios": "0.17.1",
    "babel-polyfill": "6.26.0",
    "lodash-es": "4.17.10",
    "react": "16.4.1",
    "react-dom": "16.4.1",
    "react-redux": "5.0.7",
    "react-router": "3.2.1",
    "redux": "3.7.2",
    "redux-thunk": "2.3.0"
  },
  "jest": {
    "testURL": "http://localhost",
    "testRegex": "\\.spec\\.tsx?$",
    "moduleFileExtensions": [
      "js",
      "jsx",
      "json",
      "ts",
      "tsx"
    ],
    "setupFiles": [
      "<rootDir>/config/test/polyfills.js",
      "<rootDir>/config/test/setupTest.js"
    ],
    "transform": {
      "^.+\\.(j|t)sx?$": "<rootDir>/node_modules/ts-jest/preprocessor.js"
    },
    "transformIgnorePatterns": [
      "<rootDir>/node_modules/(?!lodash-es/.*)"
    ],
    "snapshotSerializers": [
      "enzyme-to-json/serializer"
    ],
    "moduleNameMapper": {
      "^.+\\.s?(css|svg)$": "identity-obj-proxy"
    },
    "collectCoverageFrom": [
      "src/**/*.{ts,tsx}",
      "!**/node_modules/**",
      "!**/resources/**"
    ]
  }
}
