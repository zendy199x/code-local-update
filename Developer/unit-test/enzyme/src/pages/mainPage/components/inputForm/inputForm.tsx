import * as React from 'react';
import Button from '@material-ui/core/Button';
import TextField from '@material-ui/core/TextField';

interface State {
  fieldA: string;
  fieldB: string;
  fieldC: string;
}

export interface Props {
  onButtonAClick: (fieldA: string) => void;
  onButtonBClick: (fieldB: string) => void;
  onButtonCClick: (fieldC: string) => void;
}

export class InputForm extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = {
      fieldA: '',
      fieldB: '',
      fieldC: '',
    }
  }

  private handleTextFieldChange = (field: keyof State) => (event) => {
    this.setState({ [field]: event.target.value } as State);
  }

  private handleButtonClick = (field: keyof State) => () => {
    switch(field) {
      case 'fieldA':
        this.props.onButtonAClick(this.state[field]);
        break;
      case 'fieldB':
        this.props.onButtonBClick(this.state[field]);
        break;
      case 'fieldC':
        this.props.onButtonCClick(this.state[field]);
        break;
    }
  }

  public doComplicatedStuff(text?: string) {
    console.log('This method does a lof of complicated things');
    this.props.onButtonAClick(text);
    this.props.onButtonBClick(text);
    this.props.onButtonCClick(text);
    text && console.log(`Strange things are afoot: ${text}`);
    this.testingMeCanBeFun(text);
  }

  private testingMeCanBeFun = (text?: string) =>{
    if (text == null || text.length > 1) {
      console.log('Nothing to report');
    }
    switch(text) {
      case 'A':
        // Call some callbacks
        this.props.onButtonAClick('A long time ago');
        break;
      case 'B':
        // Call some more callbacks
        this.props.onButtonAClick('A long time ago');
        this.props.onButtonBClick('By the shore');
        break;
      case 'C':
        // Call even more callbacks
        this.props.onButtonAClick('A long time ago');
        this.props.onButtonBClick('By the shore');
        this.props.onButtonCClick('Close to Mordor');
        break;
      case 'D':
        // And maybe some more?
        this.props.onButtonAClick('A long time ago');
        this.props.onButtonCClick('Close to Mordor');
        this.props.onButtonCClick('Despite not being sure if Mordor has shores');
        break;
    }
  };

  public render() {
    return (
      <div>
        <TextField
          id="field_A"
          value={this.state.fieldA}
          onChange={this.handleTextFieldChange('fieldA')}
        />
        <Button
          color="primary"
          onClick={this.handleButtonClick('fieldA')}
        >
          Button A
        </Button>
        <TextField
          id="field_B"
          value={this.state.fieldB}
          onChange={this.handleTextFieldChange('fieldB')}
        />
        <Button
          color="secondary"
          onClick={this.handleButtonClick('fieldB')}
        >
          Button B
        </Button>
        <TextField
          id="field_C"
          value={this.state.fieldC}
          onChange={this.handleTextFieldChange('fieldC')}
        />
        <Button
          color="default"
          onClick={this.handleButtonClick('fieldC')}
        >
          Button C
        </Button>
      </div>
    );
  }
}
