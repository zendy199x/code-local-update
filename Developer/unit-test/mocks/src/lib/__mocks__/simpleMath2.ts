export const sum = jest.fn((a: number, b: number): any => `I'm mocked!! ${a} + ${b}`);

export const prod = jest.fn((a: number, b: number): any => `I'm mocked!! ${a} * ${b}`);

export const divide = jest.fn((a: number, b: number): any => `I'm mocked!! ${a} / ${b}`);
