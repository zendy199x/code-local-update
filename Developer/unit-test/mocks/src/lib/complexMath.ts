export const performComplexMath = (data: number[]): number => {
  if (!data || !data.length) {
    return null;
  }
  const r1 = data.reduce((prev, curr) => Math.sqrt(prev^3 + Math.log2(curr^2)), 1);
  const r2 = data.reduce((prev, curr) => Math.sqrt(curr^3 + Math.log2(prev^2)), 1);
  return r1 / r2;
};
