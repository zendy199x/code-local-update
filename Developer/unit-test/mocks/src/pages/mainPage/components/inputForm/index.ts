export { InputForm, Props as InputFormProps } from './inputForm';
