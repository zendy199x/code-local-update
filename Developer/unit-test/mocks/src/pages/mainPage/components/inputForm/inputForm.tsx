import * as React from 'react';
import Button from '@material-ui/core/Button';
import TextField from '@material-ui/core/TextField';
import { divide, sum } from '../../../../lib/simpleMath';
import { performComplexMath } from '../../../../lib/complexMath';

interface State {
  fieldA: string;
  fieldB: string;
  fieldC: string;
}

export interface Props {
  onButtonAClick: (fieldA: string) => void;
  onButtonBClick: (fieldB: string) => void;
  onButtonCClick: (fieldC: string) => void;
  generateNumberFromSeed?: (num: number) => number;
}

export class InputForm extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = {
      fieldA: '',
      fieldB: '',
      fieldC: '',
    }
  }

  private handleTextFieldChange = (field: keyof State) => (event) => {
    this.setState({ [field]: event.target.value } as State);
  }

  private handleButtonClick = (field: keyof State) => () => {
    switch(field) {
      case 'fieldA':
        this.props.onButtonAClick(this.state[field]);
        break;
      case 'fieldB':
        this.props.onButtonBClick(this.state[field]);
        break;
      case 'fieldC':
        this.props.onButtonCClick(this.state[field]);
        break;
    }
  }

  public doComplicatedStuff(text?: string) {
    console.log('This method does a lof of complicated things');
    this.props.onButtonAClick(text);
    this.props.onButtonBClick(text);
    this.props.onButtonCClick(text);
    text && console.log(`Strange things are afoot: ${text}`);
    this.testingMeCanBeFun(text);
  }

  private testingMeCanBeFun = (text?: string) =>{
    if (text == null || text.length > 1) {
      console.log('Nothing to report');
    }
    switch(text) {
      case 'A':
        // Call some callbacks
        this.props.onButtonAClick('A long time ago');
        break;
      case 'B':
        // Call some more callbacks
        this.props.onButtonAClick('A long time ago');
        this.props.onButtonBClick('By the shore');
        break;
      case 'C':
        // Call even more callbacks
        this.props.onButtonAClick('A long time ago');
        this.props.onButtonBClick('By the shore');
        this.props.onButtonCClick('Close to Mordor');
        break;
      case 'D':
        // And maybe some more?
        this.props.onButtonAClick('A long time ago');
        this.props.onButtonCClick('Close to Mordor');
        this.props.onButtonCClick('Despite not being sure if Mordor has shores');
        break;
    }
  };

  public handleMathAway = () => {
    const a = this.props.generateNumberFromSeed(Number(this.state.fieldA));
    const b = Number(this.state.fieldB);
    const c = Number(this.state.fieldC);
    const result = divide(sum(a, b), c);
    this.props.onButtonAClick(String(result));
  };

  public doSomeComplexCalculations = (data: number[]) => {
    return performComplexMath(data) + performComplexMath(data.map(d => d*d));
  }

  public render() {
    return (
      <div>
        <div>
          <TextField
            id="field_A"
            value={this.state.fieldA}
            onChange={this.handleTextFieldChange('fieldA')}
          />
          <Button
            color="primary"
            onClick={this.handleButtonClick('fieldA')}
          >
            Button A
          </Button>
          <TextField
            id="field_B"
            value={this.state.fieldB}
            onChange={this.handleTextFieldChange('fieldB')}
          />
          <Button
            color="secondary"
            onClick={this.handleButtonClick('fieldB')}
          >
            Button B
          </Button>
          <TextField
            id="field_C"
            value={this.state.fieldC}
            onChange={this.handleTextFieldChange('fieldC')}
          />
          <Button
            color="default"
            onClick={this.handleButtonClick('fieldC')}
          >
            Button C
          </Button>
        </div>
        <div>
          <Button
            id="math"
            color="primary"
            onClick={this.handleMathAway}
          >
            MATH AWAY
          </Button>
        </div>
      </div>
    );
  }
}
