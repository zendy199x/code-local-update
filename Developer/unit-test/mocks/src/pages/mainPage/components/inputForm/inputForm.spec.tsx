import * as React from 'react';
import { shallow } from 'enzyme';
import TextField from '@material-ui/core/TextField';
import Button from '@material-ui/core/Button';
import { InputForm, Props } from './inputForm';
import * as math from '../../../../lib/simpleMath';
import * as math2 from '../../../../lib/simpleMath2';
import * as complexMath from '../../../../lib/complexMath'

jest.mock('../../../../lib/simpleMath');
jest.mock('../../../../lib/simpleMath2');

const baseProps: Props = {
  onButtonAClick: () => {},
  onButtonBClick: () => {},
  onButtonCClick: () => {},
};

let props: Props = null;

describe('InputForm', () => {
  beforeEach(() => {
    props = { ...baseProps };
  });

  it('should render as expected', () => {
    const wrapper = shallow(<InputForm {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render three text fields and three buttons', () => {
    const wrapper = shallow(<InputForm {...props} />);
    expect(wrapper.find(TextField).length).toBe(3);
    expect(wrapper.find(Button).length).toBe(4);
    expect(wrapper.find(TextField).at(0).prop('id')).toBe('field_A');
    expect(wrapper.find('#field_B').prop('id')).toBe('field_B');
    expect(wrapper.find('[color="default"]').text()).toBeTruthy();
    expect(wrapper.find(Button).filter('[color="default"]')).toHaveLength(1);
  });

  it('should change the state accordingly when the user modifies the values in the text fields', () => {
    const text = 'some_text';
    const event = { target: { value: text } };
    const wrapper = shallow(<InputForm {...props} />);
    wrapper.find('#field_A').simulate('change', event);
    expect(wrapper.state('fieldA')).toBe(text);
  });

  it('should change the values in the text fields if the state is somehow changed (note: this is not a real test)', () => {
    const text = 'some_text';
    const wrapper = shallow(<InputForm {...props} />);
    wrapper.setState({ fieldA: text });
    expect(wrapper.find('#field_A').prop('value')).toBe(text);
  });

  it('should call the corresponding props callback when clicking on the buttons', () => {
    props.onButtonAClick = jest.fn();
    props.onButtonBClick = jest.fn();
    props.onButtonCClick = jest.fn();
    const wrapper = shallow(<InputForm {...props} />);
    wrapper.find(Button).at(0).simulate('click');
    wrapper.find(Button).at(1).simulate('click');
    wrapper.find(Button).at(2).simulate('click');
    expect(props.onButtonAClick).toHaveBeenCalledTimes(1);
    expect(props.onButtonBClick).toHaveBeenCalledTimes(1);
    expect(props.onButtonCClick).toHaveBeenCalledTimes(1);
  });

  it('should propagate upwards the values of field A when clicking on button A', () => {
    const expectedResult = 'value_to_propagate';
    const event = { target: { value: expectedResult } };
    props.onButtonAClick = jest.fn();
    const wrapper = shallow(<InputForm {...props} />);
    wrapper.find('#field_A').simulate('change', event);
    wrapper.find(Button).at(0).simulate('click');
    expect(props.onButtonAClick).toHaveBeenCalledWith(expectedResult);
  });

  it('should call all the callbacks when doing complicated stuff', () => {
    const text = 'So complex, much complication, wow';
    props.onButtonAClick = jest.fn();
    props.onButtonBClick = jest.fn();
    props.onButtonCClick = jest.fn();
    const wrapper = shallow(<InputForm {...props} />);
    const component = wrapper.instance() as InputForm;
    component.doComplicatedStuff(text);
    expect(props.onButtonAClick).toHaveBeenCalledWith(text);
    expect(props.onButtonBClick).toHaveBeenCalledWith(text);
    expect(props.onButtonCClick).toHaveBeenCalledWith(text);
  });

  it('should call the corresponding callbacks when sending a D to testingMeCanBeFun method', () => {
    const text = 'D';
    props.onButtonAClick = jest.fn();
    props.onButtonBClick = jest.fn();
    props.onButtonCClick = jest.fn();
    const component = new InputForm(props);
    (component as any).testingMeCanBeFun(text);
    expect(props.onButtonAClick).toHaveBeenCalledWith('A long time ago');
    expect(props.onButtonCClick).toHaveBeenCalledTimes(2);
  });
});

describe('InputForm - Mocks', () => {
  const setFields = (wrapper: any, fieldA: string, fieldB: string, fieldC: string) => {
    wrapper.find('#field_A').simulate('change', { target: { value: fieldA } });
    wrapper.find('#field_B').simulate('change', { target: { value: fieldB } });
    wrapper.find('#field_C').simulate('change', { target: { value: fieldC } });
  }

  beforeEach(() => {
    props = {...baseProps}
  })

  it('should generate a new number using field A as the seed when clicking on the Math Away button', () => {
    const [fieldA, fieldB, fieldC] = ['2', '3', '4'];
    const expectedArgument = 2;
    props.generateNumberFromSeed = jest.fn();
    props.onButtonAClick = jest.fn();
    const wrapper = shallow(<InputForm {...props} />);
    setFields(wrapper, fieldA, fieldB, fieldC);
    wrapper.find('#math').simulate('click');
    expect(props.generateNumberFromSeed).toHaveBeenCalledWith(expectedArgument);
  });

  it('should propagate the result of "math away" by calling the on-click callback for button A', () => {
    (math.sum as jest.Mock).mockReturnValueOnce(8);
    (math.divide as jest.Mock).mockReturnValueOnce(2);
    const [fieldA, fieldB, fieldC] = ['2', '3', '4'];
    const expectedResult = '2';
    const mock = jest.fn();
    mock.mockReturnValue(5);
    props.generateNumberFromSeed = mock;
    props.onButtonAClick = jest.fn();
    const wrapper = shallow(<InputForm {...props} />);
    setFields(wrapper, fieldA, fieldB, fieldC);
    wrapper.find('#math').simulate('click');
    expect(props.onButtonAClick).toHaveBeenCalledWith(expectedResult);
    expect(math.sum).toHaveBeenLastCalledWith(5, 3);
    expect(math.divide).toHaveBeenLastCalledWith(8, 4);
  });

  it('should propagate the result of "math away" as many times as it was clicked', () => {
    // Arrange
    const expectedSumCalls = [ [17, 3], [37, 3], [5, 3], [5, 3], [5, 3] ];
    const expectedDivideCalls = [ [20, 4], [40, 4], [8, 4], [8, 4], [8, 4] ];
    (math.sum as jest.Mock).mockReturnValueOnce(20);
    (math.sum as jest.Mock).mockReturnValueOnce(40);
    (math.sum as jest.Mock).mockReturnValue(8);
    (math.divide as jest.Mock).mockReturnValueOnce(5);
    (math.divide as jest.Mock).mockReturnValueOnce(10);
    (math.divide as jest.Mock).mockReturnValue(2);
    const [fieldA, fieldB, fieldC] = ['2', '3', '4'];
    const expectedCalls = [['5'], ['10'], ['2'], ['2'], ['2']];
    const generateNumberFromSeedMock = jest.fn(() => 5);
    generateNumberFromSeedMock.mockReturnValueOnce(17);
    generateNumberFromSeedMock.mockReturnValueOnce(37);
    props.generateNumberFromSeed = generateNumberFromSeedMock;
    const onButtonAClickMock = jest.fn();
    props.onButtonAClick = onButtonAClickMock;
    // Act
    const wrapper = shallow(<InputForm {...props} />);
    setFields(wrapper, fieldA, fieldB, fieldC);
    wrapper.find('#math').simulate('click');
    wrapper.find('#math').simulate('click');
    wrapper.find('#math').simulate('click');
    wrapper.find('#math').simulate('click');
    wrapper.find('#math').simulate('click');
    // Assert
    expect(onButtonAClickMock.mock.calls).toEqual(expectedCalls);
    expect((math.sum as jest.Mock).mock.calls).toEqual(expectedSumCalls);
    expect((math.divide as jest.Mock).mock.calls).toEqual(expectedDivideCalls);
  });

  it('should be calling the default manual mocks defined in __mocks__ for the simpleMath2 module imported', () => {
    const x = math2.sum(3, 5);
    const y = math2.prod(2, 7);
    const z = math2.divide(1, 9);
    expect(x).toBe(`I'm mocked!! 3 + 5`);
    expect(y).toBe(`I'm mocked!! 2 * 7`);
    expect(z).toBe(`I'm mocked!! 1 / 9`);
  });

  afterEach(() => {
    (math.sum as jest.Mock).mockReset();
    (math.divide as jest.Mock).mockReset();
  });
});

describe('InputForm - Spy', () => {
  beforeEach(() => {
    props = { ...baseProps };
  });

  it('should call performComplexMath twice when calling doSomeComplexCalculations', () => {
    const mock = jest.spyOn(complexMath, 'performComplexMath');
    const component = new InputForm(props)
    component.doSomeComplexCalculations([1, 2, 3]);
    expect(mock).toHaveBeenCalledTimes(2);
    expect(mock.mock.calls).toEqual([[[1, 2, 3]],[[1, 4, 9]]]);
    mock.mockClear();
  });

  it('should call performComplexMath twice when calling doSomeComplexCalculations v2', () => {
    const mock = jest.spyOn(complexMath, 'performComplexMath');
    mock.mockImplementation(() => 1);
    const expectedResult = 2;
    const component = new InputForm(props)
    const result = component.doSomeComplexCalculations([1, 2, 3]);
    expect(result).toBe(expectedResult);
    expect(mock).toHaveBeenCalledTimes(2);
    expect(mock.mock.calls).toEqual([[[1, 2, 3]],[[1, 4, 9]]]);
    mock.mockRestore();
  });
});