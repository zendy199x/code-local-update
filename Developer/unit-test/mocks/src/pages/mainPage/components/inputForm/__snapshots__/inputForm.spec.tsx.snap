// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputForm should render as expected 1`] = `
<div>
  <div>
    <TextField
      id="field_A"
      onChange={[Function]}
      required={false}
      select={false}
      value=""
    />
    <WithStyles(Button)
      color="primary"
      onClick={[Function]}
    >
      Button A
    </WithStyles(Button)>
    <TextField
      id="field_B"
      onChange={[Function]}
      required={false}
      select={false}
      value=""
    />
    <WithStyles(Button)
      color="secondary"
      onClick={[Function]}
    >
      Button B
    </WithStyles(Button)>
    <TextField
      id="field_C"
      onChange={[Function]}
      required={false}
      select={false}
      value=""
    />
    <WithStyles(Button)
      color="default"
      onClick={[Function]}
    >
      Button C
    </WithStyles(Button)>
  </div>
  <div>
    <WithStyles(Button)
      color="primary"
      id="math"
      onClick={[Function]}
    >
      MATH AWAY
    </WithStyles(Button)>
  </div>
</div>
`;
